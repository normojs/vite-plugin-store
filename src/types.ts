/**
 * https://github.com/brattonross/vite-plugin-voie/blob/main/packages/vite-plugin-voie/src/options.ts
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file at
 * https://github.com/brattonross/vite-plugin-voie/blob/main/LICENSE
 */

export type ImportMode = 'sync' | 'async'
export type ImportModeResolveFn = (filepath: string) => ImportMode

export interface Module {
  // 命名空间：默认开启
  namespaced?: boolean
  // 别名，默认使用文件名，允许用户自定义
  alias?: string

  // 模块名称，默认文件名或文件夹名称
  name: string
  // 文件（不包含后缀名），[name].ts，TODO index.ts、user/index.ts等
  path: string

  state: Function
  mutations: Object
  actions: Object
  getters: Object
}

/**
 *
 *
 *
    const moduleA = {
      namespaced: true,
      state: () => ({ ... }),
      mutations: { ... },
      actions: { ... },
      getters: { ... }
    }
    const store = createStore({
      strict: true,
      modules: {
        a: moduleA,
        b: moduleB
      }
    })

    interface Store 为 createStore()的参数
 */
export interface Store {
  strict: boolean // 严格模式： true（默认）|false
  namespaced?: boolean // 命名空间：默认全局开启

  name?: string // 模块名称，默认文件名或文件夹名称
  path?: string // 文件（不包含后缀名），[name].ts，TODO index.ts、user/index.ts等
  state?: Function
  mutations?: Object
  actions?: Object
  getters?: Object

  plugins?: Function[]
  // 多模块
  modules?: Module[]

}
export interface StoreDirOptions {
  dir: string
  baseRoute: string
}

export interface Route {
  name?: string
  path: string
  props?: boolean
  component: string
  children?: Route[]
  meta?: Record<string, unknown>
}

export interface PageDirOptions {
  dir: string
  baseRoute: string
}

/**
 * Plugin options.
 */
interface Options {
  /**
   * 状态组件的相对路径
   * Relative path to the directory to search for store components.
   * @default 'src/store'
   */
  storeDir: string
  /**
   * Relative path to the directory to search for page components.
   * @default 'src/pages'
   */
  pagesDir: string | string[] | PageDirOptions[]
  /**
   * Valid file extensions for page components.
   * @default ['vue', 'js']
   */
  extensions: string[]
  /**
   * List of path globs to exclude when resolving pages.
   */
  exclude: string[]
  /**
   * Import routes directly or as async components
   * @default 'async'
   */
  importMode: ImportMode | ImportModeResolveFn
  /**
   * Sync load top level index file
   * @default true
   */
  syncIndex: boolean
  /**
   * Use Nuxt.js style dynamic routing
   * @default false
   */
  nuxtStyle: boolean
  /**
   * Set default route block parser, or use `<route lang=xxx>` in SFC route block
   * @default 'json5'
   */
  routeBlockLang: 'json5' | 'json' | 'yaml'
  /**
   * Replace '[]' to '_' in bundle chunk filename
   * Experimental feature
   * @default true
   */
  replaceSquareBrackets: boolean
  /**
   * Extend route records
   */
  extendRoute?: (route: Route, parent: Route | undefined) => Route | void
}

export type UserOptions = Partial<Options>

export interface ResolvedOptions extends Options {
  /**
   * Resolves to the `root` value from Vite config.
   * @default config.root
   */
  root: string
  storeDir: string
  /**
   * Page Dir as a normalized array of PageDirOptions
   */
  pagesDirOptions: PageDirOptions[]
  /**
   * RegExp to match extensions
   */
  extensionsRE: RegExp
}
