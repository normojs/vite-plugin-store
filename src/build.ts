import { buildSync, transformSync } from 'esbuild'
export const buildModule = (fullPath: string, isTs = true) => {
  const result = buildSync({
    entryPoints: [fullPath],
    outfile: 'out.js',
    write: false,
    platform: 'browser',
    bundle: false,
    metafile: true,
    format: isTs ? 'esm' : 'cjs',
  })
  const { text } = result.outputFiles[0]
  return { text, metafile: result.metafile }
}

export const transformModule = (text: string) => {
  return transformSync(text)
}
