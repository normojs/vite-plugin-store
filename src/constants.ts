export const MODULE_IDS = ['normo:store', 'virtual:generated-store', 'store-generated']
export const MODULE_ID_VIRTUAL = '/@vite-plugin-store/generated-store'

/*
@see  https://next.vuex.vuejs.org/zh/guide/hot-reload.html
      > Vuex 支持在开发过程中热重载 mutation、module、action 和 getter
*/
export const HMR_MODULE_NAMES = ['mutations', 'actions', 'getters']
