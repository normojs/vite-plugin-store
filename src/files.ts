import fg from 'fast-glob'
import { ResolvedOptions } from './types'
import { extensionsToGlob } from './utils'

/**
 * Resolves the files that are valid pages for the given context.
 */
export async function getFilesFromPath(path: string, options: ResolvedOptions): Promise<string[]> {
  const {
    extensions,
    exclude,
  } = options

  const ext = extensionsToGlob(extensions)

  const files = await fg(`**/*.${ext}`, {
    ignore: ['node_modules', '.git', '**/__*__/*', ...exclude],
    onlyFiles: true,
    cwd: path,
  })

  return files
}
