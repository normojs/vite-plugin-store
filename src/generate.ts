/**
 * https://github.com/brattonross/vite-plugin-voie/blob/main/packages/vite-plugin-voie/src/routes.ts
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file at
 * https://github.com/brattonross/vite-plugin-voie/blob/main/LICENSE
 */

// TODO: https://www.npmjs.com/package/array-starts-with
// https://github.com/tunnckocore/starts-with
import { Route, ResolvedOptions, PageDirOptions, Store, Module, ModuleOptions, PluginOption } from './types'
import { debug, normalizePath, moduleName2InName } from './utils'
// import { stringifyStores } from './stringify'
// import { parseCustomBlock, parseSFC } from './parseSfc'

const subStoreExtensions = ['index', 'getters', 'actions', 'mutations', 'state']
type ModuleOptionsMap = Record<string, ModuleOptions>
/**
 * TODO 处理文件内容，参考: vite-plugin-components
 * 处理严格模式、工作空间
 * TODO source-map
 *
 * @export
 * @param {string[]} filesPath
 * @param {string} storeDir
 * @param {ResolvedOptions} options
 * @return {*}  {Store}
 */
export function generateOptions(filePaths: string[], storeDir: string, options: ResolvedOptions): {moduleOptions: ModuleOptions[]; pluginOptions: PluginOption[]} {
  // console.log(filesPath, storeDir, options)
  const {
    extensionsRE,
    root,
  } = options

  const moduleOptions: ModuleOptions[] = []
  const map = {}
  const moduleOptionsMap: ModuleOptionsMap = {}

  const pluginOptions: PluginOption[] = []

  for (const filePath of filePaths) {
    // 去除后缀
    const resolvedPath = filePath.replace(extensionsRE, '')
    const componentPath = `/${storeDir}/${filePath}`
    // resolvedPath: 'index' | 'user/index' | 'user/getters' | 'user/mutations'
    const temps = resolvedPath.split('/')

    // temp[0] => 'index' | 'user' | 'user' | 'user'
    let moduleName = temps[0]
    // temp[1] => | 'index' | 'getters' | 'mutations'
    let moduleInType = temps[1] || 'module'

    if (moduleName === 'plugins') {
      // 处理插件
      [moduleName, moduleInType] = [moduleInType, moduleName]
      pluginOptions.push({ resolvedPath, pluginName: moduleName, componentPath, filePath })
      continue
    }
    else if (temps.length === 1 && subStoreExtensions.includes(temps[0])) {
      // store/[actions, getters]
      moduleInType = moduleName
      moduleName = 'index'
    }
    else if (temps.length > 2) {
      // 处理多级module
      moduleInType = temps.pop() || 'undefined'
      moduleName = temps.join('/') || ''
      if (!subStoreExtensions.includes(moduleInType) || !moduleName) {
        // 如果子模块类型错误，则忽略
        continue
      }
    }

    // moduleOptionsMap

    moduleOptions.push({ root: options.root, resolvedPath, moduleName, moduleInType, componentPath, filePath })
  }// end for

  return { moduleOptions, pluginOptions }
}

/**
 * 创建一个字符串化的Vuex Store定义。
 * Store{}
 * {
 *  strict: true,
 *  state: ...
 *  getters: ...
 *  modules:{
 *
 *  }
 * }
 */
export function generateClientCode(moduleOptions: ModuleOptions[], options: ResolvedOptions) {
  const rootModule: any = { strict: true, moduleOptions, modules: {}, imports: [] }

  const moduleOptionsMap: any = {}
  moduleOptions.forEach((item: ModuleOptions) => {
    moduleOptionsMap[item.moduleName] = item
  })

  const fileSystemTree: any = {
    // 'name': result
  }
  for (const index in moduleOptions) {
    const moduleOption = moduleOptions[index]
    const moduleName: string = moduleOption.moduleName
    if (fileSystemTree[moduleName]) {
      fileSystemTree[moduleName].imports.push(moduleOption)
    }
    else {
      fileSystemTree[moduleName] = { moduleName: moduleOption.moduleName }
      fileSystemTree[moduleName].imports = [moduleOption]
    }
  }// end for

  // 【重点】 保证key的顺序
  const moduleNames = Object.keys(moduleOptionsMap).sort()
  moduleNames.reverse()

  const xx = []
  const xx2 = []
  const xx3 = []

  for (const moduleName of moduleNames) {
    const singleModule = fileSystemTree[moduleName]
    const { moduleInName, imports, modules } = generateSingleModule(singleModule, options)
    xx.push(...imports)
    xx2.push(...modules)
    xx3.push(moduleInName)

    //
    const splitKeys = moduleName.split('/')

    // 上一级module name
    let parentModuleName = ''
    if (splitKeys.length >= 2)
      parentModuleName = splitKeys.join('/')
    // TODO: 如果有parentModule，则不设置
    if (parentModuleName && fileSystemTree[parentModuleName]) {
      // 存在父module，把当前设置到父modules中
      // fileSystemTree[parentModuleName].modules || fileSystemTree[parentModuleName].modules
    }
    else {
      // 不存在父module，即：一级module
      // rootModule.modules
    }
  }

  rootModule.moduleNames = moduleNames
  rootModule.tree = fileSystemTree
  rootModule.xx = xx
  rootModule.xx2 = xx2
  rootModule.xx3 = xx3

  /*
    import _user_actions from 'user/actions'
    import _user_getters from 'user/getters'
    import _user_mutations from 'user/mutations'
    import _user_state from 'user/state'
    import _user_index from 'user/index'
    // 判断是否为空，按需拼接
    let _user_module = {
      ..._user_index,
      state: _user_state,
      mutations: ..._user_mutations,
      getters: ..._user_getters,
      actions: ..._user_actions,
      modules: {...}
    }
    */

  // TODO: 生成全部的code

  // rootModule.xx = moduleOptionsMap

  // const { name, imports } = generateModule(moduleOptionTree, options)

  // console.log('moduleOptions ；', moduleOptions)

  // return `${JSON.stringify(JSON.parse(stringStores).testString.text)}`
  return `export default ${JSON.stringify(rootModule)}`
}

interface ResultSingleModule {
  // 如：'user/role/menu' 取 'menu'
  moduleInName: string
  imports: string[]
  // TODO: 类型
  modules: string[]
  out: any
}

export function generateSingleModule(singleModule: any, options: ResolvedOptions) {
  /*
    import _user_actions from 'user/actions'
    import _user_getters from 'user/getters'
    import _user_mutations from 'user/mutations'
    import _user_state from 'user/state'
    import _user_index from 'user/index'
    // 判断是否为空，按需拼接
    let _user_module = {
      ..._user_index,
      state: _user_state,
      mutations: ..._user_mutations,
      getters: ..._user_getters,
      actions: ..._user_actions,

      modules: {
      'role': _user_role,
      'element': _user_element
      }

    }
  */
  const result: ResultSingleModule = {
    moduleInName: '',
    imports: [],
    modules: [],
    out: {},
  }
  result.moduleInName = singleModule.moduleName.split('/').pop()
  result.out.moduleVariableName = `${moduleName2InName(singleModule.moduleName)}__module`
  result.modules.push(`let _${result.out.moduleVariableName} = {`)

  singleModule.imports.forEach((item: any) => {
    const importName = moduleName2InName(`${item.moduleName}__${item.moduleInType}`)
    if (item.moduleName === 'index')
      result.modules.push(`...${importName},`)

    else if (subStoreExtensions.includes(item.moduleInType))
      result.modules.push(`${item.moduleInType}: ...${importName},`)

    result.imports.push(`import * as ${importName} from '${item.filePath}'`)
  })

  result.modules.push('}')

  return result
}

/**
 * 返回数据结构
 * TODO: 方法改名
 * @export
 * @param {string[]} paths
 * @return {*}
 * @deprecated
 */
export function pathsToTree(moduleOptions: ModuleOptions[]) {
  const moduleOptionsMap: any = {}
  moduleOptions.forEach((moduleOption) => {
    moduleOptionsMap[moduleOption.resolvedPath] = moduleOption
  })

  const result: Object[] = []
  const level = { result }
  // const subTreeMap = {}

  Object.keys(moduleOptionsMap).forEach((path) => {
    path.split('/').reduce((r: any, name, i, a) => {
      if (!r[name]) {
        r[name] = { result: [] }
        const module = moduleOptionsMap[path]
        r.result.push({ name, ...module, children: r[name].result })
      }
      return r[name]
    }, level)
  })
  return result
}

// =====================================
export function updateRouteFromHMR(content: string, filename: string, routes: Route[], options: ResolvedOptions): boolean {
  return false
}
