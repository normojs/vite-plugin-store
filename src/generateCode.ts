export const generateCode = function(root: any) {
  return `
  
  import { createStore } from 'vuex'
  // ================= 生成了store code ==================
  ${root.code}
  export const root = ${JSON.stringify({ ...root, code: '' })}
  // ================= end store code ==================

  /* 热加载 */

  // ================= 公共方法 ==================
  /**
   * 设置generatedStore下modules属性namespaced
   * @params {Object} 生成的store code，即：createStore(store)中的store
   */
  function setModulesNamespaced(moduleRoot) {
    if (!moduleRoot.modules) {
      // 结束回调
      return true
    }
    for (const moduleKey in moduleRoot.modules) {
      const module = moduleRoot.modules[moduleKey]
      if (module.namespaced === undefined) {
        // namespaced 默认值为false|undefined
        module.namespaced = true
      }
      setModulesNamespaced(module)
    }// end for
  }
  // ================= end 公共方法 ==================

  // 判断是否开启全局 namespaced，默认开启全局命名空间
  if (generatedStore.namespaced === true) {
    // 如果全局开启，则添加到模块
    setModulesNamespaced(generatedStore)
  }
  const _store = window.store
  if (!window.store) {
    console.log('不存在store, 创建generatedStore： ', generatedStore)
    window.store = createStore(generatedStore)
  }
  export const store = window.store
  export { generatedStore }

  const loadingTime = Date.now()
  const lastHotTime = loadingTime
  console.log('[vite-plugin-store] loading: ', loadingTime)

  let globalVar = 0

  let hotEvent = null
  if (import.meta.hot) {
    import.meta.hot.on('vite-plugin-store-update', (data) => {
      // TODO: 有时会有两次的更新，做防止抖动
      globalVar++
      // 判断类型
      hotEvent = data
      console.log('通知：自定义---------------------------globalVar', globalVar, loadingTime, data)
    })

    import.meta.hot.accept((newModule) => {
      console.log('操作：accept state update ...hotEvent', hotEvent, loadingTime)
      if (hotEvent) {
        const newStore = newModule.generatedStore
        console.log('newStore ', newStore)
        if (hotEvent.type === 'state') {
          // 设置state
          // TODO: 获取state，然后设置state

          let hotState = null
          // 判断是不是/index.ts
          if (hotEvent.moduleInType === hotEvent.moduleName === 'index') {
            //
            hotState = newStore.state()
          }
          else {
            // moduleName: "user/role/element"
            const subModules = hotEvent.moduleName.split('/')
            // TODO:
          }

          // TODO: 循环更新state： state => state.[user.role.menu].info,

          window.store.hotUpdate(newStore)
        }
        else if (hotEvent.type === 'hot-update') {
          window.store.hotUpdate(newStore)
        }
      }

      // console.log('newStore.modules: ', newModule, newStore)
      // let stateData = newStore.modules.account.state()

      // window.store.commit('account/__init__', stateData)

      // TODO: window.store.commit('account/__init__', stateData)
      /*
        TODO: 在handleHotUpdate方法中自动执行

        判断修改的文件路径，如果是state、index文件
        判断是否更改了state，比较出差异
        如果更改了state，则循环被修改的

        // state.user.role.menu.info
        let diffObj = {...} // 如： {'user/role/menu': {path: 'new path'}}
        window.store.commit('__hot_update_state__', diffObj)

      */
    })// end hot.accept
  }// end if

  // ============================================================

  `
}
