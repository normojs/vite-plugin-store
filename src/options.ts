import { UserOptions, ResolvedOptions, PageDirOptions } from './types'

export function resolveOptions(userOptions: UserOptions): ResolvedOptions {
  const {
    storeDir = 'src/store',
    pagesDir = ['src/pages'],
    extensions = ['ts', 'js'],
    importMode = 'async',
    routeBlockLang = 'json5',
    exclude = [],
    syncIndex = true,
    replaceSquareBrackets = false,
    nuxtStyle = false,
  } = userOptions

  const root = process.cwd()

  let pagesDirOptions: PageDirOptions[] = []

  if (typeof pagesDir === 'string') {
    pagesDirOptions = pagesDirOptions.concat({ dir: pagesDir, baseRoute: '' })
  }
  else {
    for (const dir of pagesDir) {
      if (typeof dir === 'string')
        pagesDirOptions = pagesDirOptions.concat({ dir, baseRoute: '' })
      else if (dir as PageDirOptions)
        pagesDirOptions = pagesDirOptions.concat(dir)
    }
  }

  const extensionsRE = new RegExp(`\\.(${extensions.join('|')})$`)

  return Object.assign(
    {},
    {
      routeBlockLang,
      root,
      storeDir,
      pagesDir,
      pagesDirOptions,
      extensions,
      importMode,
      exclude,
      syncIndex,
      replaceSquareBrackets,
      nuxtStyle,
      extensionsRE,
    },
    userOptions,
  )
}
