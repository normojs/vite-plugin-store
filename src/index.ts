import { resolve, basename } from 'path'
import type { Plugin, ResolvedConfig, ModuleNode } from 'vite'
import { Route, ResolvedOptions, UserOptions, ModuleOptions, Store } from './types'
import { getFilesFromPath } from './files'
import { generateModuleOptions, generateClientCode, updateRouteFromHMR } from './generate'
import { debug, normalizePath } from './utils'
import { parseVueRequest } from './query'
import { resolveOptions } from './options'
import { MODULE_IDS, MODULE_ID_VIRTUAL } from './constants'

function storePlugin(userOptions: UserOptions = {}): Plugin {
  let config: ResolvedConfig | undefined
  let filesPath: string[] = []

  let moduleOptions: ModuleOptions[]

  let generatedRoutes: Route[] | null | undefined
  let generatedStores: Store[] | null | undefined
  let generatedStore: Store | null | undefined

  const pagesDirPaths: string[] = []

  const options: ResolvedOptions = resolveOptions(userOptions)

  return {
    name: 'vite-plugin-store',
    enforce: 'pre',
    // enforce: 'post',
    configResolved(_config) {
      config = _config
      // debug.gen('config: %O', config)
      options.root = config.root
    },
    resolveId(id) {
      return MODULE_IDS.includes(id) || MODULE_IDS.some(i => id.startsWith(i))
        ? MODULE_ID_VIRTUAL
        : null
    },
    async load(id) {
      // debug.gen('load: %O', id)
      if (id && id.includes('utils/cache'))
        console.log('============:', id)

      if (id === MODULE_ID_VIRTUAL) {
        if (!generatedStores) {
          generatedStores = []
          generatedStore = { strict: true }
          filesPath = []
          const storeDirPath = normalizePath(resolve(options.root, options.storeDir))
          // 相对路径数组
          const files = await getFilesFromPath(storeDirPath, options)
          moduleOptions = generateModuleOptions(files, options.storeDir, options)
        }

        // TODO: 生成code
        const clientCode = generateClientCode(moduleOptions, options)

        return clientCode
      }
    },
    async transform(code: string, id: string) {
      const { query } = parseVueRequest(id)

      // if (id === 'vite-plugin-store')
      debug.transform('id: %O', id)
      const s = 'App.vue'
      if (id.endsWith(s))
        console.log(`code: ${s} \n`, code, '\n\n')
      // console.log('dd', id)

      // return null

      if (query && query.vue && query.type === 'route') {
        return {
          code: 'export default {}',
          map: null,
        }
      }
      return {
        code,
        // TODO source-map
        map: null,
      }
    },
    generateBundle(_options, bundle) {
      if (options.replaceSquareBrackets) {
        const files = Object.keys(bundle).map(i => basename(i))
        for (const name in bundle) {
          const chunk = bundle[name]
          chunk.fileName = chunk.fileName.replace(/(\[|\])/g, '_')
          if (chunk.type === 'chunk') {
            for (const file of files)
              chunk.code = chunk.code.replace(file, file.replace(/(\[|\])/g, '_'))
          }
        }
      }
    },
    async handleHotUpdate({ file, server, read }) {
      // TODO 热加载
      const isPagesDir = pagesDirPaths.find(p => file.startsWith(`${p}/`))
      // 是pages下的文件更新，且，
      // Handle pages HMR
      if (isPagesDir && options.extensionsRE.test(file)) {
        let needReload = false

        // Handle new file
        if (!filesPath.includes(file)) {
          generatedRoutes = null
          needReload = true
        }
        else if (generatedRoutes) {
          const content = await read()
          needReload = updateRouteFromHMR(content, file, generatedRoutes, options)
        }

        if (needReload) {
          const { moduleGraph } = server
          const module = moduleGraph.getModuleById(MODULE_ID_VIRTUAL)
          if (module)
            server.moduleGraph.invalidateModule(module)

          debug.hmr('hmr update: %s', file.replace(options.root, ''))

          return [module!]
        }
      }
    },
  }
}

export * from './types'
export default storePlugin
