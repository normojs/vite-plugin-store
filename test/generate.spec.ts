import { generateRoutes, generateClientCode } from '../src/generate'
import { resolveOptions } from '../src/options'
import { Route } from '../src/types'

const options = resolveOptions({
  pagesDir: 'test/assets/pages',
})
const nuxtOptions = resolveOptions({
  pagesDir: 'test/assets/nuxt-pages',
  nuxtStyle: true,
})
const files = ['[...all].vue', 'about.vue', 'components.vue', 'index.vue', '[sensor]/current.vue', 'about/index.vue', 'blog/[id].vue', 'blog/index.vue', 'blog/today/index.vue']
const nuxtFiles = ['_.vue', 'about.vue', 'components.vue', 'index.vue', '_sensor/current.vue', 'about/index.vue', 'blog/_id.vue', 'blog/index.vue', 'blog/today/index.vue']
const pageDir = options.pagesDirOptions[0]
const nuxtPageDir = nuxtOptions.pagesDirOptions[0]

describe('Generate', () => {
  let routes: Route[] = []
  let nuxtRoutes: Route[] = []
  test('Routes', () => {
    routes = generateRoutes(files, pageDir, options)
    expect(routes).toMatchSnapshot('routes')
  })
  test('Nuxt Style Routes', () => {
    nuxtRoutes = generateRoutes(nuxtFiles, nuxtPageDir, nuxtOptions)
    expect(nuxtRoutes).toMatchSnapshot('nuxt style routes')
  })
  test('Client Code', () => {
    const code = generateClientCode(routes, options)
    expect(code).toMatchSnapshot('client code')
  })
})
