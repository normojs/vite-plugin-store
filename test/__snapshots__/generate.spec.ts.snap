// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate Client Code: client code 1`] = `
"import _test_assets_pages_index_vue from '/test/assets/pages/index.vue'

const routes = [
  {
    \\"name\\": \\"[...all].vue\\",
    \\"path\\": \\"/[...all].vue\\",
    \\"component\\": () => import('/test/assets/pages/[...all].vue'),
    \\"props\\": true
  },
  {
    \\"name\\": \\"about.vue\\",
    \\"path\\": \\"/about.vue\\",
    \\"component\\": () => import('/test/assets/pages/about.vue'),
    \\"props\\": true
  },
  {
    \\"name\\": \\"components.vue\\",
    \\"path\\": \\"/components.vue\\",
    \\"component\\": () => import('/test/assets/pages/components.vue'),
    \\"props\\": true
  },
  {
    \\"name\\": \\"name-override\\",
    \\"path\\": \\"/index.vue\\",
    \\"component\\": _test_assets_pages_index_vue,
    \\"props\\": true,
    \\"meta\\": {
      \\"requiresAuth\\": false
    }
  },
  {
    \\"name\\": \\"sensor-current.vue\\",
    \\"path\\": \\"/:sensor/current.vue\\",
    \\"component\\": () => import('/test/assets/pages/[sensor]/current.vue'),
    \\"props\\": true
  },
  {
    \\"name\\": \\"about-index.vue\\",
    \\"path\\": \\"/about/index.vue\\",
    \\"component\\": () => import('/test/assets/pages/about/index.vue'),
    \\"props\\": true
  },
  {
    \\"name\\": \\"name-override\\",
    \\"path\\": \\"/blog/[id].vue\\",
    \\"component\\": () => import('/test/assets/pages/blog/[id].vue'),
    \\"props\\": true,
    \\"meta\\": {
      \\"requiresAuth\\": false
    }
  },
  {
    \\"name\\": \\"blog-index.vue\\",
    \\"path\\": \\"/blog/index.vue\\",
    \\"component\\": () => import('/test/assets/pages/blog/index.vue'),
    \\"props\\": true
  },
  {
    \\"name\\": \\"blog-today-index.vue\\",
    \\"path\\": \\"/blog/today/index.vue\\",
    \\"component\\": () => import('/test/assets/pages/blog/today/index.vue'),
    \\"props\\": true
  }
]

export default routes"
`;

exports[`Generate Nuxt Style Routes: nuxt style routes 1`] = `
Array [
  Object {
    "component": "/test/assets/nuxt-pages/_.vue",
    "name": ".vue",
    "path": "/:.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/nuxt-pages/about.vue",
    "name": "about.vue",
    "path": "/about.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/nuxt-pages/components.vue",
    "name": "components.vue",
    "path": "/components.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/nuxt-pages/index.vue",
    "meta": Object {
      "requiresAuth": false,
    },
    "name": "name-override",
    "path": "/index.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/nuxt-pages/_sensor/current.vue",
    "name": "sensor-current.vue",
    "path": "/:sensor/current.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/nuxt-pages/about/index.vue",
    "name": "about-index.vue",
    "path": "/about/index.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/nuxt-pages/blog/_id.vue",
    "name": "blog-id.vue",
    "path": "/blog/:id.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/nuxt-pages/blog/index.vue",
    "name": "blog-index.vue",
    "path": "/blog/index.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/nuxt-pages/blog/today/index.vue",
    "name": "blog-today-index.vue",
    "path": "/blog/today/index.vue",
    "props": true,
  },
]
`;

exports[`Generate Routes: routes 1`] = `
Array [
  Object {
    "component": "/test/assets/pages/[...all].vue",
    "name": "[...all].vue",
    "path": "/[...all].vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/pages/about.vue",
    "name": "about.vue",
    "path": "/about.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/pages/components.vue",
    "name": "components.vue",
    "path": "/components.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/pages/index.vue",
    "meta": Object {
      "requiresAuth": false,
    },
    "name": "name-override",
    "path": "/index.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/pages/[sensor]/current.vue",
    "name": "sensor-current.vue",
    "path": "/:sensor/current.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/pages/about/index.vue",
    "name": "about-index.vue",
    "path": "/about/index.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/pages/blog/[id].vue",
    "meta": Object {
      "requiresAuth": false,
    },
    "name": "name-override",
    "path": "/blog/[id].vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/pages/blog/index.vue",
    "name": "blog-index.vue",
    "path": "/blog/index.vue",
    "props": true,
  },
  Object {
    "component": "/test/assets/pages/blog/today/index.vue",
    "name": "blog-today-index.vue",
    "path": "/blog/today/index.vue",
    "props": true,
  },
]
`;
